elastic-buffer-sim/src/Bittide/Plot.hs

```diff
@@ -0,0 +1,475 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ExistentialQuantification #-}
+
+module Bittide.Plot
+  ( OutputMode(..)
+  , SimulationSettings(..)
+  , plotDiamond
+  , plotCyclic
+  , plotComplete
+  , plotGrid
+  , plotStar
+  , plotTorus2D
+  , plotTorus3D
+  , plotTree
+  , plotLine
+  , plotHyperCube
+  , simulateTopology
+  ) where
+
+import Prelude
+
+import Clash.Prelude
+  ( KnownDomain
+  , KnownNat
+  , SomeNat(..)
+  , SNat(..)
+  , type (<=)
+  , type (+)
+  , someNatVal
+  )
+
+import Clash.Sized.Vector qualified as V
+
+import System.Exit (die)
+import Text.Read (Read(..), lexP, pfail, readMaybe)
+import Text.Read.Lex (Lexeme(Ident))
+import Data.List (foldl', transpose)
+import Data.Proxy (Proxy(..))
+import Data.Aeson (ToJSON, FromJSON, Value(..))
+import Data.Aeson.Types (typeMismatch)
+import Data.Aeson qualified as A
+import Data.Text qualified as T
+import Data.Array ((!), bounds)
+import Data.Bifunctor (bimap)
+import Data.Csv (toField, encode)
+import Control.Monad (void, zipWithM_, forM_)
+import System.FilePath ((</>))
+import System.Random (randomRIO)
+import Data.ByteString.Lazy qualified as BSL (appendFile)
+
+import Data.Type.Equality ((:~:)(..))
+import GHC.TypeLits.Compare ((:<=?)(..))
+import GHC.TypeLits.Witnesses ((%<=?))
+import GHC.TypeLits.Witnesses qualified as TLW (SNat(..))
+
+import Graphics.Matplotlib
+  (Matplotlib, (%), (@@), file, plot, xlabel, ylabel, o1, o2, mp)
+
+import Bittide.Simulate (Offset)
+import Bittide.Domain (defBittideClockConfig)
+import Bittide.ClockControl (ClockControlConfig(..), clockPeriodFs)
+import Bittide.Topology (simulate, simulationEntity)
+import Bittide.Arithmetic.Ppm (diffPeriod)
+import Bittide.Topology.Graph
+
+data OutputMode =
+    CSV
+  | Plot
+  deriving (Ord, Eq)
+
+instance Show OutputMode where
+  show = \case
+    CSV  -> "csv"
+    Plot -> "plot"
+
+instance Read OutputMode where
+  readPrec = lexP >>= \case
+    Ident "csv"  -> return CSV
+    Ident "plot" -> return Plot
+    _            -> pfail
+
+instance ToJSON OutputMode where
+  toJSON = String . T.pack . show
+
+instance FromJSON OutputMode where
+  parseJSON v = case v of
+    String str -> maybe tmm return $ readMaybe $ T.unpack str
+    _          -> tmm
+   where
+    tmm = typeMismatch "OutputMode" v
+
+data SimulationSettings =
+  SimulationSettings
+    { margin    :: Int
+    , framesize :: Int
+    , points    :: Int
+    , steps     :: Int
+    , mode      :: OutputMode
+    , tmpDir    :: FilePath
+    , name      :: String
+    }
+
+plotDiamond :: SimulationSettings -> IO Bool
+plotDiamond settings@SimulationSettings{..} =
+  case ( someNat margin
+       , somePositiveNat framesize
+       ) of
+    (  Just (SomeNat (_ :: Proxy margin))
+     , Just (SomePositiveNat (_ :: Proxy framesize))
+     ) ->
+       simulateTopology defBittideClockConfig
+         (SNat @margin) (SNat @framesize) diamond settings
+    (x, y) -> invalidArgs (Just undefined, x, y)
```

Comment:
Yeah, it's ok, because `invalidArgs` only checks for `Nothing`s in the tuple, but you are right, we should get rid of the `Maybe` context first and only pass the positions to the `invalidArgs` instead.